src/Map.jl

```diff
@@ -0,0 +1,221 @@
+    return Injection(domain(g), codomain(f), f.f ∘ g.f, g.linvf ∘ f.invf)
+end
+
+function ∘(f::Injection, g::Bijection)
+    return Injection(domain(g), codomain(f), f.f ∘ g.f, g.invf ∘ f.linvf)
+end
+
+function ∘(f::Surjection, g::Surjection)
+    return Surjection(domain(g), codomain(f), f.f ∘ g.f, g.rinvf ∘ f.rinvf)
+end
+
+function ∘(f::Surjection, g::Bijection)
+    return Surjection(domain(g), codomain(f), f.f ∘ g.f, g.invf ∘ f.rinvf)
+end
+
+function ∘(f::Bijection, g::Surjection)
+    return Surjection(domain(g), codomain(f), f.f ∘ g.f, g.rinvf ∘ f.invf)
+end
+
+(m::FunctionMap)(x...) = m.f(x...)
+(m::Injection)(x...) = m.f(x...)
+(m::Bijection)(x...) = m.f(x...)
+(m::Surjection)(x...) = m.f(x...)
+
+AbstractCurve{CoD} = AbstractMap{Euclidean{Tuple{1}},CoD}
+
+AbstractField{D,T<:Tuple} = AbstractMap{D,Euclidean{T}}
+
+AbstractScalarField{D} = AbstractField{D,Tuple{1}}
+
+AbstractVectorField{D,N} = AbstractField{D,Tuple{N}}
+
+AbstractMatrixField{D,M,N} = AbstractField{D,Tuple{M,N}}
+
+struct ExponentialMap{MT,TMT,PT} <: AbstractMap{TMT,MT}
+    domain::MT
+    codomain::TMT
+    point::PT
+
+    function ExponentialMap(M::MT, x::PT) where {MT,PT}
+        # shape = representation_size(M, TVector)
+        shape = size(x)
+        TM = Euclidean(shape...)
+        return new{MT,typeof(TM),PT}(M, TM, x)
+    end
+end
+
+(m::ExponentialMap)(v) = exp(m.domain, m.point, v)
+
+show(io::IO, m::ExponentialMap) = print(io, "ExponentialMap(", m.domain, ", ", m.point, ")")
+
+
+
+
+struct LogarithmMap{TMT,MT,PT} <: AbstractMap{TMT,MT}
+    domain::TMT
+    codomain::MT
+    point::PT
+
+    function LogarithmMap(M::MT, x::PT) where {MT,PT}
+        # shape = representation_size(M, TVector)
+        shape = size(x)
+        TM = Euclidean(shape...)
+        return new{typeof(TM),MT,PT}(TM, M, x)
+    end
+end
+
+(m::LogarithmMap)(x) = log(m.codomain, m.point, x)
+
+show(io::IO, m::LogarithmMap) = print(io, "LogarithmMap(", m.codomain, ", ", m.point, ")")
+
+
+struct Geodesic{MT,ET<:ExponentialMap,VT} <: AbstractCurve{MT}
+    domain::MT
+    Exp::ET
+    tvector::VT
+end
+
+Geodesic(M, x, v) = Geodesic(M, ExponentialMap(M, x), v)
+(g::Geodesic)(t::Real) = g.Exp(t*g.tvector)
+(g::Geodesic)(T::AbstractVector) = map(t -> g.Exp(t*g.tvector), T)
+
+show(io::IO, m::Geodesic) = print(io, "Geodesic(", m.domain, ", ", m.Exp.point, ", ", m.tvector, ")")
+
+codomain(g::Geodesic) = codomain(g.Exp)
+
+
+struct ShortestGeodesic{MT,GT<:Geodesic} <: AbstractCurve{MT}
+    domain::MT
+    geodesic::GT
+end
+
+function ShortestGeodesic(M, x, y)
+    v = LogarithmMap(M, x)(y)
+    return ShortestGeodesic(M, Geodesic(M, x, v))
+end
+
+(g::ShortestGeodesic)(t) = g.geodesic(t)
+
+show(io::IO, m::ShortestGeodesic) = print(io, "ShortestGeodesic(", m.domain, ", ", m.geodesic.Exp.point, ", ", m.geodesic.tvector, ")")
+
+codomain(g::ShortestGeodesic) = codomain(g.geodesic)
+
+function derivation(γ::AbstractCurve{MT}) where {MT}
+    dim = manifold_dimension(domain(γ))
```

Comment:
You probably mean `codomain` here, right?

Reply:
Yes, that's right.